app/uk/gov/hmrc/transitmovementsrouter/connectors/PersistenceConnector.scala

```diff
@@ -69,17 +72,32 @@ class PersistenceConnectorImpl @Inject() (httpClientV2: HttpClientV2, appConfig:
   private def persistenceSendMessage(movementId: MovementId): UrlPath =
     Url(path = s"$baseRoute/traders/movements/${movementId.value}/messages").path
 
-  override def post(movementId: MovementId, messageId: MessageId, messageType: MessageType, source: Source[ByteString, _])(implicit
+  override def post(
+    movementId: MovementId,
+    messageId: MessageId,
+    messageType: MessageType,
+    source: Option[Source[ByteString, _]],
+    objectStoreURI: Option[ObjectStoreURI]
+  )(implicit
     hc: HeaderCarrier,
     ec: ExecutionContext
   ): EitherT[Future, PersistenceError, PersistenceResponse] =
     EitherT {
       val url = baseUrl.withPath(persistenceSendMessage(movementId)).withQueryString(QueryString.fromPairs("triggerId" -> messageId.value))
-      httpClientV2
+
+      val httpClient = httpClientV2
         .post(url"$url")
-        .transform(_.addHttpHeaders(HeaderNames.CONTENT_TYPE -> MimeTypes.XML, RouterHeaderNames.MESSAGE_TYPE -> messageType.code))
-        .withBody(source)
-        .execute[Either[UpstreamErrorResponse, HttpResponse]]
+      ((source, objectStoreURI) match {
+        case (Some(src), _) =>
+          httpClient.transform(_.addHttpHeaders(HeaderNames.CONTENT_TYPE -> MimeTypes.XML, RouterHeaderNames.MESSAGE_TYPE -> messageType.code)).withBody(src)
+        case (_, Some(objectStoreURIHeader)) =>
+          httpClient.transform(
+            _.addHttpHeaders(
+              RouterHeaderNames.MESSAGE_TYPE     -> messageType.code,
+              RouterHeaderNames.OBJECT_STORE_URI -> objectStoreURIHeader.value
+            )
+          )
```

Comment:
So what happens when you have neither or both? Your function suggests that's possible when it shouldn't be. Just because we don't call it that way, doesn't mean that someone in live services who takes over in a years' time will understand that.

This is too complicated for what it needs to be. While I'm a big fan of "Don't Repeat Yourself", I'm also a fan of keeping things simple. Really, you're calling this endpoint in two very distinct ways.

Why not make two functions like so:

```
  def postBody(
    movementId: MovementId,
    messageId: MessageId,
    messageType: MessageType,
    source: Source[ByteString, _]
  )(implicit
    hc: HeaderCarrier,
    ec: ExecutionContext
  ): EitherT[Future, PersistenceError, PersistenceResponse]

  def postObjectStoreUri(
    movementId: MovementId,
    messageId: MessageId,
    messageType: MessageType,
    objectStoreURI: ObjectStoreURI
  )(implicit
    hc: HeaderCarrier,
    ec: ExecutionContext
  ): EitherT[Future, PersistenceError, PersistenceResponse]
```

an in each case set up the client. Creating the URL and the http client could be done in its own function too if you wanted. Once that's done, you can have a third method that takes the client and then does the common execution part. That way it's clear:

* These are two different calls for two different cases
* You must have one of the body, or the object store URL
* The response is the same regardless

No point shoehorning all this into one function if it makes this less clear.

Reply:
Done the changes as suggested above. Please re-review.